addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java

```diff
@@ -0,0 +1,403 @@
+            cookieManager.put(connection.getURL().toURI(), connection.getHeaderFields());
+            response = IOUtils.toString(connection.getInputStream());
+        } else if (rc == HTTP_UNAUTHORIZED && !loginGuard) {
+            // our session has timed-out, let's try again
+            if (login()) {
+                return get(url, params, true); // guard against a "login loop"
+            }
+        }
+        return response;
+    }
+
+    private String get(String url, Map<String, String> params) throws IOException, URISyntaxException {
+        return get(url, params, false);
+    }
+
+    private String get(String url) throws IOException, URISyntaxException {
+        return get(url, (Map<String, String>) null);
+    }
+
+    private <T> T get(String url, Map<String, String> params, Class<T> responseType) throws UniFiException {
+        T response = null;
+        try {
+            String content = get(url, params);
+            if (StringUtils.isBlank(content)) {
+                logger.warn("Empty GET response from URL: {}", url);
+                throw new UniFiException("Could not communicate with the UniFi Controller");
+            } else {
+                if (logger.isTraceEnabled()) {
+                    logger.trace("JSON content from URL: {}\n{}", url, objectMapper.writerWithDefaultPrettyPrinter()
+                            .writeValueAsString(objectMapper.readValue(content, Object.class)));
+                }
+                response = objectMapper.readValue(content, responseType);
+            }
+        } catch (SocketException | SocketTimeoutException e) {
+            throw new UniFiException("Could not communicate with the UniFi Controller");
+        } catch (Exception e) {
+            logger.error("Error GET'ing URL: {}", url, e);
+            throw new UniFiException(e);
+        }
+        return response;
+    }
+
+    private <T> T get(String url, Class<T> responseType) throws UniFiException {
+        return get(url, null, responseType);
+    }
+
+    public boolean login() {
+        boolean success = false;
+        // login
+        Map<String, String> params = new HashMap<String, String>();
+        params.put("username", config.getUsername());
+        params.put("password", config.getPassword());
+        try {
+            String response = get(getLoginUrl(), params);
+            success = (response != null);
+        } catch (SocketException | SocketTimeoutException e) {
+            logger.warn("Could not connect to the UniFi Controller");
+            success = false;
+        } catch (Exception e) {
+            logger.error("Error trying to authenticate to UniFi Controller", e);
+            success = false;
+        }
+        return success;
+    }
+
+    private Map<String, UniFiSite> getSites() throws UniFiException {
+        UniFiSiteResponse response = get(getSitesUrl(), UniFiSiteResponse.class);
+        Map<String, UniFiSite> siteMap = new HashMap<String, UniFiSite>();
```

Comment:
Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

Please check all